src/types/struct.rs

```diff
@@ -57,12 +63,130 @@ pub fn build<'ctx>(
+                {
+                    needs_override = true;
+                    break;
+                }
+            }
+
+            needs_override
+                .then(|| build_dup(context, module, registry, metadata, &info))
+                .transpose()
+        },
+    )?;
+    DropOverridesMeta::register_with(
+        context,
+        module,
+        registry,
+        metadata,
+        info.self_ty(),
+        |metadata| {
+            // The following unwrap is unreachable because `register_with` will always insert it
+            // before calling this closure.
+            let mut needs_override = false;
+            for member in &info.members {
+                registry.build_type(context, module, registry, metadata, member)?;
+                if metadata
+                    .get::<DropOverridesMeta>()
+                    .unwrap()
+                    .is_overriden(member)
+                {
+                    needs_override = true;
+                    break;
+                }
+            }
+
+            needs_override
+                .then(|| build_drop(context, module, registry, metadata, &info))
+                .transpose()
+        },
+    )?;
+
+    let members = info
         .members
         .iter()
-        .map(|field| registry.build_type(context, module, registry, metadata, field))
-        .collect::<Result<_>>()?;
-    let struct_ty = llvm::r#type::r#struct(context, &fields, false);
+        .map(|member| registry.build_type(context, module, registry, metadata, member))
+        .collect::<Result<Vec<_>>>()?;
+    Ok(llvm::r#type::r#struct(context, &members, false))
+}
+
+fn build_dup<'ctx>(
+    context: &'ctx Context,
+    module: &Module<'ctx>,
+    registry: &ProgramRegistry<CoreType, CoreLibfunc>,
+    metadata: &mut MetadataStorage,
+    info: &WithSelf<StructConcreteType>,
+) -> Result<Region<'ctx>> {
+    let location = Location::unknown(context);
+
+    let self_ty = registry.build_type(context, module, registry, metadata, info.self_ty())?;
+
+    let region = Region::new();
+    let entry = region.append_block(Block::new(&[(self_ty, location)]));
+
+    let mut src_value = entry.argument(0)?.into();
+    let mut dst_value = entry.append_op_result(llvm::undef(self_ty, location))?;
+
+    for (idx, member_id) in info.members.iter().enumerate() {
+        let member_ty = registry.build_type(context, module, registry, metadata, member_id)?;
+        let member_val = entry.extract_value(context, location, src_value, member_ty, idx)?;
+
+        // The following unwrap is unreachable because the registration logic will always insert it.
+        let values = metadata
+            .get::<DupOverridesMeta>()
+            .unwrap()
+            .invoke_override(context, &entry, location, member_id, member_val)?;
+
+        src_value = entry.insert_value(context, location, src_value, values.0, idx)?;
+        dst_value = entry.insert_value(context, location, dst_value, values.1, idx)?;
+    }
```

Comment:
This will only dup members that have an override, while if a member is copyable, i won't be cloned to the new struct. Am I right, or I'm missing something?

Reply:
This snippet will:
  - For each member in the struct (regardless of whether it's copyable or requires clone):
    - Build its type (makes sure the clone implementation is registered.
    - Extract the value.
    - Invoke the clone function. If registered, then it'll be cloned. Otherwise the same value is returned twice (therefore copying it).
    - Store the results into the original and cloned structs. The original is also stored just in case we need it in the future.
    - Update `src_value` and `dst_value` variables so that the next member builds from them.

This means that it'll work for both copyable and cloneable members. I'm not sure if this answers your question.

Reply:
Aaah, I see! Thanks for the explanation